block/provider/create_snapshot.go

```diff
@@ -18,61 +18,85 @@
 package provider
 
 import (
+	"time"
+
+	"github.com/IBM/ibmcloud-volume-interface/lib/metrics"
 	"github.com/IBM/ibmcloud-volume-interface/lib/provider"
+	"github.com/IBM/ibmcloud-volume-interface/lib/utils/reasoncode"
 	userError "github.com/IBM/ibmcloud-volume-vpc/common/messages"
 	"github.com/IBM/ibmcloud-volume-vpc/common/vpcclient/models"
 	"go.uber.org/zap"
 )
 
-// CreateSnapshot Create snapshot from given volume
-func (vpcs *VPCSession) CreateSnapshot(volumeRequest *provider.Volume, tags map[string]string) (*provider.Snapshot, error) {
-	vpcs.Logger.Info("Entry CreateSnapshot", zap.Reflect("volumeRequest", volumeRequest))
-	defer vpcs.Logger.Info("Exit CreateSnapshot", zap.Reflect("volumeRequest", volumeRequest))
-
-	if volumeRequest == nil {
-		return nil, userError.GetUserError("StorageFindFailedWithVolumeId", nil, "Not a valid volume ID")
-	}
+const snapshotReadyState = "stable"
 
-	var snapshot *models.Snapshot
+// CreateSnapshot creates snapshot
+func (vpcs *VPCSession) CreateSnapshot(sourceVolumeID string, snapshotParameters provider.SnapshotParameters) (*provider.Snapshot, error) {
+	vpcs.Logger.Info("Entry CreateSnapshot", zap.Reflect("snapshotRequest", snapshotParameters), zap.Reflect("sourceVolumeID", sourceVolumeID))
+	defer vpcs.Logger.Info("Exit CreateSnapshot", zap.Reflect("snapshotRequest", snapshotParameters), zap.Reflect("sourceVolumeID", sourceVolumeID))
+	defer metrics.UpdateDurationFromStart(vpcs.Logger, "CreateSnapshot", time.Now())
 	var err error
 
+	vpcs.Logger.Info("Validating basic inputs for CreateSnapshot method...", zap.Reflect("snapshotRequest", snapshotParameters), zap.Reflect("sourceVolumeID", sourceVolumeID))
+	err = vpcs.validateSnapshotRequest(sourceVolumeID, snapshotParameters)
+	if err != nil {
+		return nil, err
+	}
+	var snapshotResult *models.Snapshot
+
 	// Step 1- validate input which are required
-	vpcs.Logger.Info("Requested volume is:", zap.Reflect("Volume", volumeRequest))
+	vpcs.Logger.Info("Requested volume is:", zap.Reflect("Volume", sourceVolumeID))
 
 	var volume *models.Volume
 	err = retry(vpcs.Logger, func() error {
-		volume, err = vpcs.Apiclient.VolumeService().GetVolume(volumeRequest.VolumeID, vpcs.Logger)
+		volume, err = vpcs.Apiclient.VolumeService().GetVolume(sourceVolumeID, vpcs.Logger)
```

Comment:
Not required to get volume before taking snapshot. Rely on snapshot api for response

Reply:
check snapshot present with same name is present or not, if present then return error from here itself and not call vpc api